niflow_manager/cli/build.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+import subprocess as sp
+import tempfile, os, json
+from pathlib import Path
+from copy import deepcopy
+import yaml
+import click
+import neurodocker as ndr
+
+# default setting for specific neurodocker keys,
+# this value will not have to be set in the parameters.yml, but can be overwritten
+DEFAULT_INSTRUCTIONS = {"miniconda": {"create_env": "testkraken", "activate": True}}
+# all keys allowed by neurodocker for Docker
+VALID_DOCKER_KEYS = ndr.Dockerfile._implementations.keys()
+
+
+def neurodocker_dict(workflow_path):
+    """reading spec.yml and creating neurodocker dict
+        for the environment defined in the build part
+    """
+    with (workflow_path / "spec.yml").open() as f:
+        params = yaml.safe_load(f)
+    instructions = []
+
+    if "build" not in params:
+        raise Exception("spec.yml file has to contain the build part")
+    if "required_env" not in params["build"]:
+        raise Exception("the build part of the spec has to contain required_env")
+    # treating required_env as one env spec
+    params_env = params["build"]["required_env"]
+    # checking if entrypoint is provided
+    params_entry = params["build"].get("entrypoint", None)
+
+    if "base" not in params_env.keys():
+        raise ValueError("base image has to be provided")
+
+    miniconda_env = None
+    for key, spec in params_env.items():
+        if key == "base":
+            base_image = spec.get("image", None)
+            if base_image is None:
+                raise Exception("image has to be provided in base")
+            this_instruction = ("base", base_image)
+            pkg_manager = spec.get("pkg_manager", None)
+            if pkg_manager is None:
+                if base_image in ["centos", "fedora"]:
+                    pkg_manager = "yum"
+                else:
+                    pkg_manager = "apt"  # assume apt
+        elif key in VALID_DOCKER_KEYS:
+            spec_final = deepcopy(DEFAULT_INSTRUCTIONS.get(key, {}))
+            spec_final.update(spec)
+            this_instruction = (key, spec_final)
+            if key == "miniconda":
+                miniconda_env = spec_final["create_env"]
+        else:
+            raise Exception(
+                f"{key} is not a valid key, must be "
+                f"from the list {VALID_DOCKER_KEYS}"
+            )
+        instructions.append(this_instruction)
+    # TODO will be removed
+    instructions.insert(1, ("install", ["git"]))
+
+    # adding post build part
+    post_build = params.get("post_build", {})
+    if not post_build:
+        # if post build not provided, it will use the default one that installs niflow-manager
+        # and the package (after coping it first)
+        post_build["copy"] = [".", "/nfm"]
+        # TODO should be updated and the git could be removed
+        post_build["miniconda"] = {
+            "pip_install": [
+                "https://github.com/djarecka/niflow-manager/tarball/new_testkraken"
```

Comment:
why should `testkraken` be installed in just the build process? `nfm build` should only build the container for running.

also this path seems to be very specific.

Reply:
also, say my workflow is a py27 or even py35 workflow. a niflow should be able to support such a workflow. testkraken because of its pydra dependency is going to be py37+. so there has to be a difference between the python workflow environment and the test environment.

Reply:
this does not install `testkraken`, but `niflow-mamanger`. For now I'm installing from my branch (the one is used for the pr), otherwise it will not work...

Reply:
i think you could  point to master, which should work when this is merged. 

but we should think of a way to do this bit differently so it can refer to a version somehow.

Reply:
I opened a new issue: https://github.com/niflows/niflow-manager/issues/29

i agree, it's important to think about it more